src/compiler/parser.ts

```diff
@@ -2818,7 +2821,23 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    // Make sure we *do* perform ASI for constructs like this:
+                    //    var x = foo
+                    //    as (Bar)
+                    // This should be parsed as an initialized variable, followed
+                    // by a function call to 'as' with the argument 'Bar'
+                    if (canParseSemicolon()) {
+                        break;
+                    }
+                    else {
+                        nextToken();
+                        leftOperand = makeAsExpression(leftOperand, parseType());

```

Comment:
There is an ambiguity with union types worth calling out here:
```ts
x as Foo | Bar
```
This is intended to be a type assertion to a union type, and not a bitwise OR where the first operand is a type assertion. Seems like you do the right thing here, but just wanted to bring it up.